src/Provider/ReturnTypeProvider/CollectionFirstAndLast.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Weirdan\DoctrinePsalmPlugin\Provider\ReturnTypeProvider;
+
+use Psalm\Plugin\EventHandler\Event\MethodReturnTypeProviderEvent;
+use Psalm\Plugin\EventHandler\MethodReturnTypeProviderInterface;
+use Psalm\Type;
+use function is_string;
+
+class CollectionFirstAndLast implements MethodReturnTypeProviderInterface
+{
+    public static function getClassLikeNames() : array
+    {
+        return ['Collection'];
+    }
+
+    public static function getMethodReturnType(MethodReturnTypeProviderEvent $event) : ?Type\Union
+    {
+        if (
+            $event->getMethodNameLowercase() !== 'first'
+            || $event->getMethodNameLowercase() !== 'last'
+        ) {
+            return null;
+        }
+
+        $stmt = $event->getStmt();
+        if (!$stmt instanceof \PhpParser\Node\Expr\MethodCall) {
+            return null;
+        }
+
+        if (!$stmt->var instanceof \PhpParser\Node\Expr\Variable) {
+            return null;
+        }
+
+        if (!is_string($stmt->var->name)) {
+            return null;
+        }
+
+        $scopedVarName = '$' . $stmt->var->name . '->isEmpty()';
+        if (!isset($event->getContext()->vars_in_scope[$scopedVarName])) {
+            return null;
+        }
+
+        $type = $event->getContext()->vars_in_scope[$scopedVarName];
+
+        if ($type->isFalse()) {
+            return new Type\Union([
+                new Type\Atomic\TGenericObject()
+            ]);
```

Comment:
I don't know how to filter false here

Reply:
I assume tests fails with this code? I can't see the CI results. What's the error when you run it with this code?

Reply:
It's not even valid PHP since the constructor of TGenericObject has two required parameters.
But the Collection could have anything inside, so I'm not sure I should generate this by myself.

I would have expect doing something like
```
return TypeCombinator::remove($event->getReturnType(), new NullType())
```
without knowing what is the real syntax to
- Access the default return type of the method
- Removing null from it

Reply:
You should be able to fetch the type with `$statements_source->getNodeTypeProvider()->getType($stmt);`
Then, `$type->removeType('bool');` should be enought

Reply:
Thanks.

I suppose I have to add my `MethodReturnTypeProvider` class somehow in the `Plugin` class ? 
Do you know how I add it ?

Reply:
here's an example: https://github.com/orklah/psalm-insane-comparison/blob/master/Plugin.php

Reply:
So I added 
```
$psalm->registerHooksFromClass(CollectionFirstAndLast::class);
```
but my test is still considering `false` as a possible value. Any idea why ?

Even if I add a `var_dump` or a `die` in my class (or even in the Plugin class), it has no impact...